### PR TITLE
Test: Adding integration tests for cv_device_v3

### PR DESCRIPTION
**Pull request by Vibhu-gslab**
## Change Summary

Adding integration tests for cv_device_v3

## Related Issue(s)

Fixes #<ISSUE ID>

## Component(s) name

`arista.cvp.cv_device_v3`

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly. (check the box if not applicable)
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/cvp/examples/inventory.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+---
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/examples/inventory.yml]: Can you remove this file from this PR? It will be added by another PR, that will be merged in before this

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        parentContainerName: "Spine"
+
+    CVP_DEVICES_PARENT_CONTAINER_NONE:
+      - fqdn: leaf1
+        parentContainerName: None
+
+    CVP_DEVICES_PARENT_CONTAINER_SAME:
+      - fqdn: leaf1
+        parentContainerName: "Spine"
+    CVP_DEVICES_STATE_ABSENT:
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: new line

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 3]
```diff
@@ -0,0 +1,95 @@
+---
+- name: Test cv_device_v3
+  hosts: CloudVision
+  connection: local
+  gather_facts: no
+  vars:
+    CVP_DEVICES_MOVE_DEVICE:
+      - fqdn: leaf1
+        parentContainerName: "Spine"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: I think you should move this to `Leaf` container just to make things logical

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    CVP_DEVICES_PARENT_CONTAINER_NONE:
+      - fqdn: leaf1
+        parentContainerName: None
+
+    CVP_DEVICES_PARENT_CONTAINER_SAME:
+      - fqdn: leaf1
+        parentContainerName: "Spine"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: Change to `Leaf`

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  - name: Run CV_DEVICE_V3 To Move Device
+    arista.cvp.cv_device_v3:
+        devices: '{{CVP_DEVICES_MOVE_DEVICE}}'
+        state: present
+    register: MOVE_DEVICE_RUN
+  - assert:
+      that:
+        - "MOVE_DEVICE_RUN.devices_moved.changed == true"
+        - "MOVE_DEVICE_RUN.devices_moved.devices_moved_count == 1"
+        - "MOVE_DEVICE_RUN.devices_moved.devices_moved_list == ['leaf1_to_Spine']"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: `leaf1_to_Leaf`

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        - MOVE_DEVICE.taskIds: []
+
+#####################################
+#         STATE ABSENT              #
+#####################################
+  - name: Run CV_DEVICE_V3 With State Absent
+    arista.cvp.cv_device_v3:
+      devices: '{{CVP_DEVICES_STATE_ABSENT}}'
+      state: absent
+    #    ignore_errors: yes # ignoring errors so assertions can be done
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: remove comment line

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 2 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: added

File: ansible_collections/arista/cvp/examples/inventory.yml (hunk 1 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/examples/inventory.yml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 3 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 4 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 5 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 6 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+    CVP_DEVICES_PROVISIONING_RESET:
+      - fqdn: leaf3
+        parentContainerName: ""
+
+    CVP_DEVICES_BUNDLE_ATTACHED:
+      - serialNumber: JPE504a004ea054
+        parentContainerName: L2_Leaf
+        configlets:
+          - 'AVD_Ipmi3'
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: Can we use a custom configlet and not what we assume to already be present on ATD?

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+        imageBundle: leaf_bundle
+
+    CVP_DEVICES_BUNDLE_DETACHED:
+      - serialNumber: JPE504a004ea054
+        parentContainerName: L2_Leaf
+        configlets:
+          - 'AVD_Ipmi3'
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: Same comment as above

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+          - 'configlet1'
+          - 'configlet2'
+
+    CVP_DEVICES_DETACH_CONFIGLET:
+      - fqdn: leaf1
+        parentContainerName: "Leaf"
+
+    CVP_DEVICES_DEPLOY:
+      - fqdn: leaf3    # leaf1 must be in undefined container
+        parentContainerName: "pod2"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: This container doesn't exist on ATD. Please create container and then deploy it there.

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 10]
```diff
@@ -0,0 +1,308 @@
+---
+- name: Test cv_device_v3
+  hosts: CloudVision
+  connection: local
+  gather_facts: no
+  vars:
+    CVP_DEVICES_MOVE_DEVICE:
+      - fqdn: leaf1
+        parentContainerName: "Leaf"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: Let's create a new container and then move it there.

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+      - fqdn: leaf2
+        parentContainerName: ""
+
+    CVP_DEVICES_PROVISIONING_RESET:
+      - fqdn: leaf3
+        parentContainerName: ""
+
+    CVP_DEVICES_BUNDLE_ATTACHED:
+      - serialNumber: JPE504a004ea054
+        parentContainerName: L2_Leaf
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: Did we create this container or is it something that is assumed to exist on ATD by default?

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        configlets:
+          - 'AVD_Ipmi3'
+        imageBundle: leaf_bundle
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: Can we upload an image and then use it here instead of using something that's already present on ATD?

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+    CVP_CONFIGLETS:
+      configlet1: '! This is first configlet'
+      configlet2: '! This is second configlet'
+
+    CVP_DEVICES_APPLY_CONFIGLET:
+      - fqdn: leaf1
+        parentContainerName: "Leaf"
+        configlets:
+          - 'configlet1'
+          - 'configlet2'
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: upload these configlets before attaching please

**Comment by sugetha24** [comment]
Please upload images, configlets etc and then use it in the test rather than using something that is assumed to be present on ATD.

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_move_and_deploy_device.yml [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+    - name: Pause for 10 seconds to Execute Task
+      ansible.builtin.pause:
+        seconds: 10
+
+    - name: Validate configurations
+      arista.cvp.cv_validate_v3:
+        devices: "{{CVP_DEVICES_VALIDATE_CONFIG}}"
+        validate_mode: stop_on_error
+      register: CVP_DEVICES_RESULTS
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_move_and_deploy_device.yml:133]: Why are we validating the configs here? We have molecule tests for cv_validate_v3 separately.

Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_move_and_deploy_device.yml:133]: Tamas suggested to verify the configlets before applying it in deploy_device.
We were facing the issue of device getting inactive if the configlets are not correct.

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 13 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 12 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 11 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 10 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 9 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 8 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml (hunk 7 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_cv_device_v3.yaml]: fixed

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        action: add
+        image: "{{IMAGE_NAME}}"
+
+    - name: "Create bundle {{inventory_hostname}}"
+      arista.cvp.cv_image_v3:
+        mode: bundle
+        action: add
+        bundle_name: "{{BUNDLE_NAME}}"
+        image_list:
+          - "{{IMAGE_NAME}}"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml:53]: You need a Terminattr to be added to the bundle. Please add a terminattr to this bundle

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+  tasks:
+    - name: Collect devices facts from {{inventory_hostname}}
+      arista.cvp.cv_facts_v3:
+        facts:
+          - devices
+        regexp_filter: "{{DEVICE_NAME}}"
+      register: CV_FACTS_V3_RESULT
+
+    - name: "Push config"
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml]: Why are we testing push configlet here rather than in `test_apply_detach_configlets.yaml` ?

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_decommission_factory_reset_provisioning_reset.yml [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+    ######################################
+    - name: Run CV_DEVICE_V3 With State Absent
+      arista.cvp.cv_device_v3:
+        devices: '{{CVP_DEVICES_STATE_ABSENT}}'
+        state: absent
+      register: DECOMMISSION_DEVICE
+    - assert:
+        that:
+          - DECOMMISSION_DEVICE.devices_decommissioned.changed == true
+          - DECOMMISSION_DEVICE.devices_decommissioned.devices_decommissioned_count == 0
```
Comment (sugetha24) [inline ansible_collections/arista/cvp/molecule/cv_device_v3/test_decommission_factory_reset_provisioning_reset.yml]: The count here should be 1

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml (hunk 15 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml:70]: done

Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml:53]: done

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_decommission_factory_reset_provisioning_reset.yml (hunk 17 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_decommission_factory_reset_provisioning_reset.yml]: updated this already!
yesterday I tested with lab

File: ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml (hunk 16 above)
Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml]: we are not testing it here, just uploading it to be used while attaching/detaching bundle

Comment (Shivani-chourasiya) [reply ansible_collections/arista/cvp/molecule/cv_device_v3/test_apply_detach_bundle.yml]: Tested it without configlets, removing configlet part.
Thanks!

**Review by sugetha24 — APPROVED** [review]

**Review by noredistribution — APPROVED** [review]
tested successfully on dual DC ATD, lgtm!+1
